src/i18n.ts

```diff
@@ -0,0 +1,342 @@
+
+	/**
+	 * An optional method that is capable of owning handles, presumably to destroy them when the context object
+	 * is destroyed.
+	 *
+	 * @param handle
+	 * A handle that, when destroyed, will prevent the context object from being invalidated when the locale
+	 * is changed.
+	 *
+	 * @return A handle to *unown* the passed handle.
+	 */
+	own?: (handle: Handle) => Handle;
+}
+
+interface LocaleModule<T extends Messages> {
+	default?: T;
+}
+
+/**
+ * An object of keys to locale messages.
+ */
+export interface Messages {
+	[key: string]: string;
+}
+
+const PATH_SEPARATOR: string = has('host-node') ? global.require('path').sep : '/';
+const VALID_PATH_PATTERN = new RegExp(PATH_SEPARATOR + '[^' + PATH_SEPARATOR + ']+$');
+const contextObjects: LocaleContext[] = [];
+let rootLocale: string;
+
+/**
+ * @private
+ * Load the specified locale-specific bundles, mapping the default exports to simple `Messages` objects.
+ */
+const loadLocaleBundles = (function () {
+	function mapMessages<T extends Messages>(modules: LocaleModule<T>[]): T[] {
+		return modules.map(function (localeModule: LocaleModule<T>): T {
+			return localeModule.default as T;
+		});
+	}
+
+	if (has('host-node')) {
+		return function<T extends Messages>(paths: string[]): Promise<T[]> {
+			const modules = paths.map(function (path: string): LocaleModule<T> {
+				return global.require(path) as LocaleModule<T>;
+			});
+
+			return Promise.all(mapMessages(modules));
+		};
+	}
+
+	return function<T extends Messages>(paths: string[]): Promise<T[]> {
+		return new Promise<T[]>(function (resolve, reject) {
+			const require = global.require;
+			require.on('error', function (error: Error) {
+				reject(error);
+			});
+			require(paths, function (...modules: LocaleModule<T>[]) {
+				resolve(mapMessages(modules));
+			});
+		});
+	};
+})();
+
+/**
+ * @private
+ * Retrieve a list of supported locales that can provide messages for the specified locale.
+ *
+ * @param locale
+ * The target locale.
+ *
+ * @param supported
+ * The locales that are supported by the bundle.
+ *
+ * @return
+ * A list of supported locales that match the target locale.
+ */
+function getSupportedLocales(locale: string, supported: string[] = []): string[] {
+	const normalized = normalizeLocale(locale);
+	const parts = normalized.split('-');
+	const result: string[] = [];
+
+	let current = parts[0];
+
+	if (supported.indexOf(current) > -1) {
+		result.push(current);
+	}
+
+	for (let i = 0; i < parts.length - 1; i += 1) {
+		const next = parts[i + 1];
+
+		if (next) {
+			current += '-' + next;
+
+			if (supported.indexOf(current) > -1) {
+				result.push(current);
+			}
+		}
+	}
+
+	return result;
+}
+
+/**
+ * @private
+ * A type guard that determines whether a context object is stateful and can be invalidated.
+ *
+ * @param value
+ * The target value.
+ */
+function isContextObject(value: any): boolean {
+	return value.state && typeof value.own === 'function' && typeof value.invalidate === 'function';
```

Comment:
Might be worth wrapping this in a `Boolean(value.state && typeof value.own === 'function' && typeof value.invalidate === 'function')`